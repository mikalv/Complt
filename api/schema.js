import {
  GraphQLSchema,
  GraphQLObjectType,
  GraphQLID,
  GraphQLNonNull,
  GraphQLList,
} from 'graphql';
import Item from './types/Item';
import Task from './types/Task';
import Project from './types/Project';
import User from './types/User';
import TaskInput from './inputs/TaskInput';
import TaskUpdateInput from './inputs/TaskUpdateInput';
import ProjectInput from './inputs/ProjectInput';
import getItemsByUserAndIds from './lib/getItemsByUserAndIds';
import addItemToProject from './lib/addItemToProject';
import verifyItemExists from './lib/verifyItemExists';
import updateTask from './lib/updateTask';
import deleteTask from './lib/deleteTask';

const schema = new GraphQLSchema({
  query: new GraphQLObjectType({
    name: 'Query',
    fields: {
      itemById: {
        type: Item,
        args: {
          id: { type: new GraphQLNonNull(GraphQLID) },
        },
        resolve(rootValue, { id }) {
          if (id !== 'inbox' && id !== 'root') {
            return getItemsByUserAndIds(rootValue.userId, [id]).then(items => items[0]);
          }
          return new Error(`Please do not request the ${id} project from this field`);
        },
      },
      inbox: {
        name: 'Inbox',
        type: new GraphQLList(Task),
        resolve(rootValue) {
          return verifyItemExists(rootValue.userId, 'inbox')
            .then(inbox => getItemsByUserAndIds(rootValue.userId, inbox.children));
        },
      },
      root: {
        name: 'Root',
        type: new GraphQLList(Item),
        resolve(rootValue) {
          return verifyItemExists(rootValue.userId, 'root')
            .then(rootProject => getItemsByUserAndIds(rootValue.userId, rootProject.children));
        },
      },
      user: {
        name: 'User',
        type: User,
        resolve: ({ user }) => user,
      },
    },
  }),
  mutation: new GraphQLObjectType({
    name: 'Mutation',
    fields: {
      taskUpdate: {
        args: {
          input: { type: TaskUpdateInput },
        },
        type: Task,
        resolve({ userId }, { input }) {
          return updateTask(userId, input);
        },
      },
      deleteTask: {
        args: {
          parentProjectId: { type: new GraphQLNonNull(GraphQLID) },
          taskId: { type: new GraphQLNonNull(GraphQLID) },
        },
        type: Task,
        resolve({ userId }, { parentProjectId, taskId }) {
          return deleteTask(userId, taskId, parentProjectId);
        },
      },
      createProject: {
        args: {
          projectId: { type: new GraphQLNonNull(GraphQLID) },
          project: { type: new GraphQLNonNull(ProjectInput) },
        },
        type: Project,
        resolve({ userId }, { project, projectId }) {
          const projectToCreate = {
            name: project.name,
            projectType: project.isSequential ? 'seq' : 'para',
            owner: userId,
            isProject: true,
          };
          if (projectId === 'inbox') {
            return new Error('Projects cannot be added to the inbox');
          }
          return addItemToProject(projectToCreate, projectId)
          .then(projectCreated => projectCreated);
        },
      },
      createTask: {
        args: {
          projectId: { type: new GraphQLNonNull(GraphQLID) },
          task: { type: new GraphQLNonNull(TaskInput) },
        },
        type: Task,
        resolve({ userId }, { task, projectId }) {
          const taskToCreate = {
            name: task.name,
            isCompleted: task.isCompleted,
            tags: task.tags,
            owner: userId,
            isProject: false,
          };
          return addItemToProject(taskToCreate, projectId).then(taskCreated => taskCreated);
        },
      },
    },
  }),
});

export default schema;
