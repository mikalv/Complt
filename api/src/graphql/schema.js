import {
  GraphQLSchema,
  GraphQLObjectType,
  GraphQLID,
  GraphQLNonNull,
  GraphQLList,
} from 'graphql';
import Item from './types/Item';
import Task from './types/Task';
import TaskInput from './inputs/TaskInput';
import TaskUpdateInput from './inputs/TaskUpdateInput';
import getItemsByUserAndIds from '../lib/getItemsByUserAndIds';
import { verifyInboxExists, addItemToInbox } from '../lib/inbox';
import updateTask from '../lib/updateTask';

const schema = new GraphQLSchema({
  query: new GraphQLObjectType({
    name: 'Query',
    fields: {
      itemById: {
        type: Item,
        args: {
          id: { type: new GraphQLNonNull(GraphQLID) },
        },
        resolve(rootValue, { id }) {
          if (id !== 'inbox' && id !== 'root') {
            return getItemsByUserAndIds(rootValue.userId, [id]).then(items => items[0]);
          }
          return new Error(`Please do not request the ${id} project from this field`);
        },
      },
      inbox: {
        name: 'Inbox',
        type: new GraphQLList(Task),
        resolve(rootValue) {
          return verifyInboxExists(rootValue.userId)
            .then(inbox => getItemsByUserAndIds(rootValue.userId, inbox.children));
        },
      },
    },
  }),
  mutation: new GraphQLObjectType({
    name: 'Mutation',
    fields: {
      createTaskInInbox: {
        args: {
          input: { type: TaskInput },
        },
        type: Task,
        resolve({ userId }, { input }) {
          return addItemToInbox({
            ...input,
            owner: userId,
            isProject: false,
          }).then(task => task);
        },
      },
      taskUpdate: {
        args: {
          input: { type: TaskUpdateInput },
        },
        type: Task,
        resolve({ userId }, { input }) {
          return updateTask(userId, input);
        },
      },
    },
  }),
});

export default schema;
