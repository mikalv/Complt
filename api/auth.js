import authenticationClient from './authClient';

function getToken(tokenWithBearer) {
  const match = tokenWithBearer.match(/^Bearer (.*)$/);
  if (!match || match.length < 2) {
    return null;
  }
  return match[1];
}

const getUserId = (tokenWithBearer) => {
  const token = getToken(tokenWithBearer);
  if (token === null) return Promise.reject('Token not found');
  return authenticationClient.tokens.getInfo(token).then((user) => {
    if (typeof user === 'object') return user;
    return Promise.reject('Unauthorized');
  });
};

export default getUserId;
