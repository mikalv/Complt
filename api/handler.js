import { graphql } from 'graphql';
import schema from './src/graphql/schema';
import getUserId from './src/auth';

module.exports.graphql = (event, context, callback) => {
  getUserId(event.headers.Authorization).then((user) => {
    if (user === 'Unauthorized') {
      const response = {
        statusCode: 401,
        headers: {
          'Access-Control-Allow-Origin': '*',
        },
        body: 'Unauthorized',
      };
      callback(null, response);
    } else {
      const body = JSON.parse(event.body);
      graphql(schema, body.query, { userId: user.user_id }, undefined, body.variables, body.operationName).then((result) => {
        const response = {
          statusCode: 200,
          headers: {
            'Access-Control-Allow-Origin': '*',
          },
          body: JSON.stringify(result),
        };
        callback(null, response);
      });
    }
  });
};
