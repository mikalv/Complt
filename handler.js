import { graphql } from 'graphql';
import schema from './api/schema';
import getUserId from './api/auth';

module.exports.graphql = (event, context, callback) => {
  getUserId(event.headers.Authorization).then((user) => {
    if (user === 'Unauthorized' || user === 'Not enough or too many segments') {
      const response = {
        statusCode: 401,
        headers: {
          'Access-Control-Allow-Origin': '*',
        },
        body: 'Unauthorized',
      };
      callback(null, response);
    } else {
      const body = JSON.parse(event.body);
      graphql(schema, body.query, { userId: user.user_id, user }, undefined, body.variables,
        body.operationName).then((result) => {
          const response = {
            statusCode: 200,
            headers: {
              'Access-Control-Allow-Origin': '*',
            },
            body: JSON.stringify(result),
          };
          callback(null, response);
        });
    }
  });
};
