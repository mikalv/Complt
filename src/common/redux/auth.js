import { LOGIN, LOGOUT } from './actionTypes';

export const initialState = '';

export default function authReducer(state = initialState, action) {
  switch (action.type) {
    case LOGIN:
      return action.token;
    case LOGOUT:
      return initialState;
    default:
      return state;
  }
}
