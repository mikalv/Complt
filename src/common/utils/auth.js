function decodeJWT(jwt) {
  const array = jwt.split('.');
  if (array.length !== 3) return;
  const data = JSON.parse(global.atob(array[1]));
  return data; // eslint-disable-line consistent-return
}

const getTokenExpirationDate = token => {
  try {
    const decoded = decodeJWT(token);
    if (!decoded || !decoded.exp) {
      return null;
    }
    const date = new Date(0);
    date.setUTCSeconds(decoded.exp);
    return date;
  } catch (error) {
    return null;
  }
};

const isTokenExpired = token => {
  if (!token) return true;
  const date = getTokenExpirationDate(token);
  if (date === null) return true;
  return !(date.valueOf() > new Date().valueOf());
};

export default isTokenExpired;
