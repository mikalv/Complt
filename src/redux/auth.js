import { LOGIN, LOGOUT } from './actionTypes';

export const initialState = {
  token: {
    idToken: '',
  },
};

export default function authReducer(state = initialState, action) {
  switch (action.type) {
    case LOGIN:
      localStorage.setItem('token', action.token.idToken);
      return {
        ...state,
        token: action.token,
      };
    case LOGOUT:
      localStorage.removeItem('token');
      return initialState;
    default:
      return state;
  }
}
