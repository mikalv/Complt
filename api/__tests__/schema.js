import schema from '../schema';
import '../auth';
import '../../handler';

jest.mock('auth0');
// Currently this test is only here so all code is included in coverage reports
describe('GraphQL schema', () => {
  it('is an object', () => {
    expect(typeof schema).toEqual('object');
  });
});
