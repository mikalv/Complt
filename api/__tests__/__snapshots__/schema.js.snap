exports[`GraphQL schema matches the snapshot 1`] = `
"union Item = Task | Project

type Mutation {
  taskUpdate(input: TaskUpdateInput): Task
  deleteTask(parentProjectId: ID!, taskId: ID!): Task
  deleteProject(parentProjectId: ID!, projectId: ID!): Project
  createProject(projectId: ID!, project: ProjectInput!): Project
  createTask(projectId: ID!, task: TaskInput!): Task
}

# A project that can contain tasks or other projects
type Project {
  id: ID!
  name: String
  isSequential: Boolean
  children: [Item]
}

input ProjectInput {
  name: String!
  isSequential: Boolean!
}

type Query {
  itemById(id: ID!): Item
  inbox: [Task]
  root: [Item]
  user: User
}

# A task
type Task {
  id: ID!
  name: String
  isCompleted: Boolean
  tags: [String]
}

input TaskInput {
  name: String!
  isCompleted: Boolean!
  tags: [String]
}

input TaskUpdateInput {
  id: ID!
  name: String
  isCompleted: Boolean
  tags: [String]
}

# The data of the user who is performing the request
type User {
  name: String
  email: String
  picture: String
}
"
`;
